### ISSUE TITLE
Ignoring merge commits—can we remove `--no-merges`?

### ISSUE THREAD

**Issue by OliverJAsh** [GoogleChrome/lighthouse-ci#691]
We just finished integrating Lighthouse CI and server into the CI process at Unsplash and we quickly noticed weird behaviour with merge commits.

For example, here is a PR with a normal commit (`Clarify`) immediately followed by a merge commit where we merged `master` into the branch to bring the branch up-to-date with `master`. If we inspect the merge commit we can see that Lighthouse ran as part of the checks:

![](https://user-images.githubusercontent.com/921609/133470347-b654baf1-1a1e-4c79-8d80-02fe0ab89b67.png)

However, it is missing the status check. Surprisingly, the status check was posted to the previous commit (`Clarify`) rather than the merge commit:

![](https://user-images.githubusercontent.com/921609/133470362-afb49e81-ea93-46a6-ba56-f14100d2fcd8.png)

Furthermore, when we inspect the generated Lighthouse report that was uploaded to the server, it also uses the previous commit (`Clarify`) rather than the merge commit.

After searching around I discovered this is semi-intentional, as described here: https://github.com/GoogleChrome/lighthouse-ci/pull/186#issuecomment-575688693.

> Several CI providers automatically merge master into the branch before building which means if we considered merges the hash that we report would be one that doesn't exist anywhere else and is relatively meaningless, we only want to report a hash that will exist for the user elsewhere.

I completely understand. In fact GitHub workflows/actions does this exactly. However, I don't think this is a reason to ignore _intentional_ merge commits that are actually part of the branch.

Take this git history for example:

```
A - B - C - D - E - F - G (master)
          \           \
            1 - 2 - 3 - 4 (feature)
```

Commit 4 is an _intentional_ merge commit where we merged `master` into the `feature` branch. We run the Lighthouse check for this commit, but the Lighthouse report and status will say that it ran against commit 3. This is incorrect and misleading: commit 4 and commit 3 could be substantially different, because commit 4 includes changes from `master` which could affect the Lighthouse score etc.

For this reason I believe the Lighthouse report and status should use the commit that the check actually ran against, i.e. commit 4. In order for this to happen we would need to remove `--no-merges`. We should still ignore those ephemeral merge commits that are created by CI providers, but this should be done in other ways, e.g. with GitHub actions we checkout the correct original commit: https://github.com/GoogleChrome/lighthouse-ci/issues/172#issuecomment-571928200

```yml
- uses: actions/checkout@v2
  with:
    ref: ${{ github.event.pull_request.head.sha }}
```

**Comment by patrickhulce** [comment]
Thanks for filing @OliverJAsh!

I'm hesitant to remove the `--no-merges` *by default*. You are of course correct there are others ways of going about it using the CI configuration, but in practice we have found that most users do no follow our docs verbatim and tend to copy only snippets that they believe are relevant, so this behavior came about *by default* because the failure scenario in that case is LHCI runs cannot be found *at all* instead of just on the commit before the merge. 

My preference here would be expand the coverage of CI environments that work correctly out of the box (i.e. use `GITHUB_HEAD_REF` for GitHub Actions) and to add the section of `env: LHCI_BUILD_CONTEXT__CURRENT_HASH` to the docs for any `pull_request` examples.

**Comment by uncaught** [comment]
I would agree this to be a bug. Because the current retrieval `git rev-list --no-merges -n1 HEAD` will **not** return the relevant commit after an actual merge. The merge can have integrated much older commits and thus that resulting hash is simply wrong and not event part of the merge.

I like the idea to hide the meaningless temporary merge commits in a pull/merge request, but that's definitely wrong for normal branches. I would prefer the default to be inconvenient rather than wrong though.

GitLab can run the pipeline for merge requests with and without merging it with the target branch. So all of this even depends on the project settings and can't be anticipated by the script.

For GitLab this works because `CI_MERGE_REQUEST_SOURCE_BRANCH_SHA` is empty if this is not a merge request:

```bash
export LHCI_BUILD_CONTEXT__CURRENT_HASH=${CI_MERGE_REQUEST_SOURCE_BRANCH_SHA:-$CI_COMMIT_SHA}
```
